NXP Add SCMI P2A workflow and BBNSM protocol

yongxu-wang15 opened the pull request:
1. SCMI P2A support: Adds RX channel handling to support notifications from platform to agent, including callback categorization for different message types.
2. Shared memory separation: Extends SCMI shared memory to an phandles to allow TX and RX to use different memory regions, which is required on platforms like i.MX95.
3. BBNSM protocol support: Adds initial support for the BBNSM domain protocol
test pass in imx95 m7 hello_word sample, The M7 core can continuously receive button messages from the SCMI platform.

yongxu-wang15 committed:
drivers: firmware: scmi: Add support for SCMI P2A workflow

SCMI P2A (Platform to Agent) communication is critical in platform-agent
interactions. Previously, only A2P (Agent to Platform) was supported.
This patch introduces the P2A workflow by adding support for
`scmi_channel->rx` and `scmi_mbox_channel->rx`, treating the RX channel
as an independent channel.

For P2A RX channels, initialization is optional and depends on platform
design. If both the protocol node and the base transport node do not
specify an RX shmem region, the RX channel will not be initialized.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
dts: bindings: scmi: Update shmem from phandle into phandles

Extends the SCMI shared memory structure to phandles,
allowing TX and RX channels to use different shared memory regions.
This is necessary for platforms like i.MX95, where TX and RX
notifications are mapped to different addresses.

RX channel is not mandatory and depends on the platform's support

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
soc: nxp: imx95: enable scmi bbm button notification for m7 core

enable scmi bbm button function, m7 can get notification from
scmi platform when press onoff button

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented:
v2: updated to fix twister build issue in imx95/imx943 m/a core

dbaluta commented on drivers/firmware/scmi/core.c:
```diff
@@ -176,6 +171,62 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+		ret = k_mutex_lock(&proto->rx->lock, K_USEC(SCMI_CHAN_LOCK_TIMEOUT_USEC));
+		if (ret < 0) {
+			LOG_ERR("failed to acquire chan lock");
+			return ret;
```
Here you are returning with mutex locked. I think you might want to `goto out_release_mutex`.

dbaluta commented on drivers/firmware/scmi/core.c:
```diff
@@ -62,11 +62,6 @@ static int scmi_core_setup_chan(const struct device *transport,
 		return 0;
 	}
 
```
Here you are mixing protocol generic stuff e.g introduction of `SCMI P2A` with specific handlers like for BBM. Please separate them in different patches.

yongxu-wang15 replied:
yes, have remove it into another commit

dbaluta commented on doc/hardware/arch/arm-scmi.rst:
```diff
@@ -191,6 +191,17 @@ example, ``CPU_SLEEP_MODE_SET``.
 	This driver is NXP-specific. As such, it may only be used on NXP
 	system that uses SCMI for cpu domain management operations.
 
+NXP - BBNSM management
```
At least in the commit message you should explain what BBNSM acronym stands for so that you can allow reviewers to understand what is this all about.

yongxu-wang15 replied:
sure, updated it

yongxu-wang15 commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -6,9 +6,58 @@
+
+#if defined(CONFIG_NXP_SCMI_BBM_HELPERS)
+	uint32_t flags;
+
```
have removed this call back flow into another commit, thanks

dleach02 commented on drivers/firmware/scmi/shmem.c:
```diff
@@ -58,6 +63,31 @@ __weak int scmi_shmem_vendor_write_message(struct scmi_shmem_layout *layout)
+	const struct scmi_shmem_config *cfg;
+
+	data = shmem->data;
+	cfg = shmem->config;
```
@yongxu-wang15 the AI review is flagging a legit issue here. The "cfg" is not referenced after this assignment. Static analysis will flag this later.

dleach02 reviewed (dismissed):
please clean up the unused cfg assignment

yongxu-wang15 commented:
Updated: 
1. @dleach02 thanks review, removed this unused variable
2.  rebased into newly commit to get scmi_send_message changed in another pr, and add same use_polling parameter into scmi_bbm_button_notify
3. read message don't need pre_kernel flow now, only use interrupt mode in scmi_read_message api.

thanks a lot

azizidomar-arm commented on dts/bindings/firmware/arm,scmi.yaml:
```diff
@@ -47,11 +47,11 @@ compatible: "arm,scmi"
     description: |
-      Phandle to node describing TX channel shared memory area.
+      Phandle to node describing TX/RX channel shared memory area.
       This translates to a **single** SCMI transmit channel.
```
Maybe update the description, as this will translate to tx and rx channels?

azizidomar-arm commented on drivers/firmware/scmi/core.c:
```diff
@@ -226,16 +271,26 @@ static int scmi_core_protocol_setup(const struct device *transport)
 #ifndef CONFIG_ARM_SCMI_TRANSPORT_HAS_STATIC_CHANNELS
 		/* no static channel allocation, attempt dynamic binding */
 		it->tx = scmi_transport_request_channel(transport, it->id, true);
+		it->rx = scmi_transport_request_channel(transport, it->id, true);
```
Should be false as we're requesting a RX channel

yongxu-wang15 replied:
sure, updated it

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/protocol.h:
```diff
@@ -34,6 +34,15 @@
+ * @brief SCMI header extraction
+ *
+ */
+#define SCMI_HEADER_MSG_EX(header)  (((header) & 0xFFU) >> 0U)
```
sure, unified SCMI_MESSAGE_HDR_MAKE and SCMI_MESSAGE_HDR_EX now

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -216,6 +211,56 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+	}
+
+	/* send p2a response into platform */
+	ret = scmi_transport_send_message(proto->transport, proto->rx, msg);
```
yes, for basic scmi protocol, agent don't need send reponse into platform after get p2a message, have removed it, only reserved free channel status code flow which is needed, thanks

yongxu-wang15 commented:
Updated: modified PR from @azizidomar-arm comments
1. put rx channel as optional which depend on dts defined
2. p2a not need resonse, reset channel status after got msg is enough.
3. unified SCMI_MESSAGE_HDR_MAKE and SCMI_HEADER_MSG_EX
4. separate tx and notification rx callback
5. change imx95 m7 to support rx, not enabled it in other platform

thanks a lot!

yongxu-wang15 commented:
Updated：
fixed twister issue.

Thanks

LaurentiuM1234 reviewed (commented):
Left some initial comments below.

The first issue that I've noticed with the proposed design is that it mixes the protocol and
transport layers which should be kept independent. Therefore, we need to rethink the way
we handle notifications.

Perhaps each protocol could keep a list of message IDs and callbacks to be invoked by the core whenever the platform sends a notification? Note that the SCMI core acts as a glue between the protocol and transport layers and, therefore, you have a bit more freedom with what you can do there.

yongxu-wang15 commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -6,21 +6,120 @@
+	protocol_id = SCMI_MESSAGE_HDR_EX_PROTOCOL(msg.hdr);
+	msg_id = SCMI_MESSAGE_HDR_EX_MSGID(msg.hdr);
+
+#if defined(CONFIG_NXP_SCMI_BBM_HELPERS)
```
removed it now

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/shmem.h:
```diff
@@ -93,4 +95,24 @@ int scmi_shmem_vendor_write_message(struct scmi_shmem_layout *layout);
+ * @retval 0 if successful
+ * @retval negative errno if failure
+ */
+int scmi_shmem_read_hdr(const struct device *shmem, struct scmi_message *msg);
```
sure, done

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -236,6 +257,18 @@ static int scmi_core_protocol_setup(const struct device *transport)
 		}
+
+		/*
+		 * rx is optional and can be either NULL or resolved to
```
ok, done

yongxu-wang15 commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -6,21 +6,120 @@
 
 #include <zephyr/logging/log.h>
 #include "mailbox.h"
+#if defined(CONFIG_NXP_SCMI_BBM_HELPERS)
```
removed it from mailbox, thanks

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -216,6 +211,31 @@ int scmi_send_message(struct scmi_protocol *proto, struct scmi_message *msg,
+	 * required in the PRE_KERNEL stage. The interrupt-based logic is used here.
+	 */
+	ret = scmi_transport_read_message(proto->transport, proto->rx, msg);
+	if (ret < 0) {
```
sure, resolved it now, thanks

yongxu-wang15 commented:
Updated PR from @LaurentiuM1234  @dleach02 
1. separate core layer and protocol level for notification event, protocol side take register event into core layer list, such as bbm button event, and the core layer maintains a list of registered protocol event handlers and dispatches incoming P2A notifications to the appropriate protocol based on the protocol ID and message ID.
2. fixed some code style issue and no quality issue found by ci now.

Thanks a lot

yongxu-wang15 commented:
Hi, Guys,

It has been modified as suggested, any comments about it?

Thanks

yongxu-wang15 commented:
Updated: removed not reasonable LOG info

Thanks

yongxu-wang15 commented:
Hi @LaurentiuM1234  , please help to review this pr again when you time free, thanks

dbaluta commented:
@yongxu-wang15 Looks good to me.

LaurentiuM1234 commented on drivers/firmware/scmi/core.c:
```diff
@@ -15,6 +15,8 @@ LOG_MODULE_REGISTER(scmi_core);
 #define SCMI_CHAN_LOCK_TIMEOUT_USEC 500
 #define SCMI_CHAN_SEM_TIMEOUT_USEC 500
 
+static sys_slist_t scmi_event_list;
```
Do we really need a linked list? Can we not have each protocol keep an array of `struct scmi_protocol_event` for all registered events?

The definition of `struct scmi_protocol_event` would be:

```
struct scmi_protocol_event {
    uint32_t id; /* event ID */
    scmi_evt_callback_t cb; /* event callback */
};
```

LaurentiuM1234 commented on drivers/firmware/scmi/nxp/bbm.c:
```diff
@@ -67,3 +74,33 @@ int scmi_bbm_button_event(uint32_t *flags)
+
+static int scmi_nxp_bbm_event_init(void)
+{
+	return scmi_register_protocol_event_handler(&bbm_event);
```
do we need to also unregister protocol events? if not, then we can probably get around with binding the event array to the protocol structure at build time.

LaurentiuM1234 commented on dts/bindings/firmware/arm,scmi.yaml:
```diff
@@ -48,11 +48,13 @@ include: [base.yaml]
 properties:
   shmem:
-    type: phandle
+    type: phandle-array
```
can this not be phandles type?

LaurentiuM1234 reviewed (changes requested):
Some questions/comments below.

None of your commits should break the build if applied incrementally. If I try to build the `hello_world` application with 6fbb947cc908 and then 9a3dda1216db applied, the build will break.

I believe @dbaluta has told you multiple times to break your changes into different patches, based on what the changes are trying to achieve. This, of course, applies to the whole PR, not just the areas on which @dbaluta commented.

As things stand now I don't believe this is ready for merge, which is why I'm blocking this.

yongxu-wang15 committed:
drivers: firmware: scmi: add event registration and core dispatcher

Add support for protocol-level registration of notification events
during static registration. Each protocol declares the notifications
it needs at static init time.

In the SCMI core layer, upon receiving a P2A notification, the core
iterates over all registered protocols to locate the matching event
and dispatches it to the appropriate protocol handler.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add NXP bbnsm domain protocol

This protocol is intended provide access to the battery-backed module,
it contains persistent storage (GPR), an RTC, and the ON/OFF button.
The protocol can also provide access to similar functions
implemented via external board components.

Added button P2A notification event for the BBM protocol.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
dts: arm: nxp: nxp_imx95_m7: add scmi rx shmem and bbm node

Add scmi rx shmem and bbm node into imx95 m7

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add scmi_read_message for notification handling

Introduce scmi_read_message() to allow protocols to read the associated
event payload when a P2A notification is received. This enables protocol
drivers to process notification-specific data after the SCMI core
dispatches the event.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: Rename mailbox tx callback name

Rename scmi_mbox_cb to scmi_mbox_tx_reply_cb, this is a preparation step
for adding dedicated RX channel support.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: Add RX channel callback support in mailbox

Implement dedicated RX channel support with separate notification callback.
TX and RX channels now use different callbacks for proper message handling.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on include/zephyr/drivers/firmware/scmi/util.h:
```diff
@@ -97,22 +107,28 @@
 /**
- * @brief Get a reference to a protocol's SCMI TX channel
+ * @brief Get a reference to a protocol's SCMI TX/RX channel
  *
```
sure, done

yongxu-wang15 commented on drivers/firmware/scmi/core.c:
```diff
@@ -55,6 +83,10 @@ static void scmi_core_reply_cb(struct scmi_channel *chan, int hdr)
+		status = scmi_core_handle_notification(hdr);
+		if (status) {
+			LOG_WRN("Scmi notification event not find");
+		}
```
yes, I think its needed to add it in there

dbaluta commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -10,9 +10,21 @@
 LOG_MODULE_REGISTER(scmi_mbox);
 
 static void scmi_mbox_tx_reply_cb(const struct device *mbox,
```
Am I confused by github diff? But it looks like scmi_mbox_tx_reply_cb and scmi_mbox_notify_cb are identical?

Also, why don't you use scmi_mbox_notify_rx_cb instead of scmi_mbox_notify_cb?

yongxu-wang15 replied:
not identical between tx with rx notification callback, we need to clear channel in rx callback, tx is not need, and use different callback to diff tx and rx is more clarifier

dbaluta commented on include/zephyr/drivers/firmware/scmi/util.h:
```diff
@@ -305,6 +305,16 @@
+ */
+#define SCMI_FIELD_EX(x, mask, shift) \
+	((((uint32_t)(x)) >> (shift)) & (uint32_t)(mask))
+
```
If to set the fileld the name is SCMI_FIELD_MAKE, please use SCMI_FIELD_TAKE for the opposite operation.

yongxu-wang15 replied:
yes, done

dbaluta commented:
@yongxu-wang15 mostly looking good! I've made few comments. You can ping me internally when need for a new review.

yongxu-wang15 committed:
drivers: firmware: scmi: add SCMI_FIELD_TAKE macro

Add SCMI_FIELD_TAKE macro, This will be called at the driver level

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add interface to read hdr message

Introduced an interface to shmem to directly read the current
hdr content without knowing the context of the message

Add api to get shmem message in mailbox driver

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add msg_type validation in callback

In the SCMI core layer, add msg_type validation before dispatch
(e.g. RESP vs NOTIFY vs CMD). Messages with an unexpected or invalid
type for the current context are logged

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -27,9 +27,13 @@ static void scmi_mbox_notify_cb(const struct device *mbox,
+	struct scmi_mbox_channel *mbox_chan = scmi_chan->data;
+	const struct device *shmem = mbox_chan->shmem;
+	bool a2p = false;
 
```
thanks comment about this, there are some error in previous commit about clear channel status, updated now.
 and tx  callback not need to set/clear channel status, sender will mark channel as busy and proceed, then another cpu side will to deal with the status

dbaluta commented on drivers/firmware/scmi/shmem.c:
```diff
@@ -39,6 +39,19 @@ int scmi_shmem_get_channel_status(const struct device *dev, uint32_t *status)
+	layout->chan_status |= SCMI_SHMEM_CHAN_STATUS_FREE_BIT;
+
+	return 0;
+}
```
Looks this function can be void. No need to always return 0.

yongxu-wang15 replied:
yes, done

yongxu-wang15 committed:
drivers: firmware: scmi: Rename SCMI_SHMEM_CHAN_STATUS_BUSY_BIT

Rename SCMI_SHMEM_CHAN_STATUS_BUSY_BIT to SCMI_SHMEM_CHAN_STATUS_FREE_BIT
for clearer semantics (bit set = channel free, bit clear = channel busy)

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: add api to clear channel status

Add scmi_shmem_clear_channel_status() function that always sets
channel to FREE state after message processing.
It provides a standardized way for higher layer components
(such as mailbox drivers) to explicitly clear the SCMI shared
memory channel status, ensuring proper channel state management
and avoiding potential race conditions in multi-layer SCMI
implementations.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 committed:
drivers: firmware: scmi: clear rx channel status in rx notification cb

Clear the SCMI shared memory channel status after processing RX
notifications in the mailbox callback for proper P2A flow management.

In SCMI P2A (Platform to Application) flow, the application side
must explicitly clear the channel status after processing incoming
messages from the platform. This is different from A2P (Application
to Platform) TX flow where the cleanup is handled by the peer
(platform) side.

Signed-off-by: Yongxu Wang <yongxu.wang@nxp.com>

yongxu-wang15 commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -27,9 +27,12 @@ static void scmi_mbox_rx_notify_cb(const struct device *mbox,
 				struct mbox_msg *data)
```
have split it into 3 commit now, firstly commit to rename SCMI_SHMEM_CHAN_STATUS_BUSY_BIT, then add api to set channel free, finally apply this function in mailbox rx notification callback, this will be clearer, thanks

azizidomar-arm commented on drivers/firmware/scmi/core.c:
```diff
@@ -43,8 +43,55 @@ int scmi_status_to_errno(int scmi_status)
+	}
+
 	if (!k_is_pre_kernel()) {
 		k_sem_give(&chan->sem);
```
Why give the semaphore here? It should be given only when msg_type is SCMI_COMMAND, no?

yongxu-wang15 replied:
you are right, it only need by SCMI_COMMAND, updated it now, thanks

azizidomar-arm commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -6,18 +6,60 @@
+
+	msg->hdr = 0x0;
+	msg->len = sizeof(uint32_t);
+	msg->content = &context;
```
Pointer to a stack variable? Should be NULL instead, right?

azizidomar-arm commented on drivers/firmware/scmi/mailbox.c:
```diff
@@ -6,18 +6,60 @@
-			 mbox_channel_id_t channel_id,
-			 void *user_data,
-			 struct mbox_msg *data)
+static int scmi_mbox_get_pending_msg(struct scmi_channel *chan,
```
I’m not a fan of the function name since it can be misleading. What do you think about changing it to something like scmi_mbox_read_msg_hdr?

yongxu-wang15 commented:
Updated:
Rebased into main to fix code conflict as newly added other scmi code, no other code change compared with previous
test log as this:
<img width="1208" height="190" alt="image" src="https://github.com/user-attachments/assets/cd094321-10f7-491f-90d9-42dbfce342cd" />

mmahadevan108 commented:
@LaurentiuM1234 , please re-review. I have made you the assignee.
